### PR TITLE
Core service structure

### PR DESCRIPTION
**Pull request by achilleas-k**
Initial PR with core structure.  It's not all there yet, but it's enough to get started.

## Overview
Overview of the layout
- **Tonic** service: The top level of the service.  All implementations will be instances of this struct with the `ProcessForm` function and form elements defined separately for each.
  - **web.Server**: The web server component.
  - **Database**: A sqlite database that holds the submitted jobs.
    - **JobInfo**: Maps to the database table for the jobs and holds all information required for a job, minus the actual action performed.
  - **Worker**: Maintains a job queue and runs them in order of submission.  The actual task carried out by each job is implemented separately for each instance.  This can be defined inside the ProcessForm implementation of the **Tonic** service, which will have access to each submitted form's data and can bind the values through a closure.
    - **Job**: Extends the ORMed **JobInfo** with the function carried out for each job.
  - **Form**: The form is currently defined as a list of **Element** instances.  I'm planning to extend this to support multiple pages of elements.

## Notes and plans
- Web routes are defined outside the web layer at the **Tonic** service layer.  This is useful for having access to the database or the job queue as needed.  I also intend to add the ability for implementations to define extra routes in case an implementation requires one.
- The configuration struct is there but there's no way to get data into it yet.
- I intend to add a top-level logger that's attached to the **Tonic** service.  Implementations can use the same logger for their own messaging.
- I will also add a session store in the database.  Currently, the user's token is stored in a cookie.  Instead, I want to store a session ID that maps to a cookie in the database as well as any other information we might need.
- I expect more details will change in the next PR when I finalise the implementation for the research project creation bot.  I'll also add a bot that does nothing except create empty jobs that print a log line as an example and test implementation.

### CODE REVIEW

File: tonic/db/job.go [hunk 1]
```diff
@@ -0,0 +1,75 @@
...
+	if err := conn.engine.Find(&userjobs, condition); err != nil {
+		return nil, err
+	}
+
+	return userjobs, nil
+}
+
+// IsFinished returns true if the Job has finished (has an EndTime).
+func (ji *JobInfo) IsFinished() bool {
+	return ji.EndTime.IsZero()
```
Comment (mpsonntag) [inline tonic/db/job.go:52]: The description above states '0' if job is ongoing. Does this then not return true if the job is still running?

File: tonic/routes.go [hunk 2]
```diff
@@ -0,0 +1,161 @@
...
+
+	tmpl := template.New("layout")
+	tmpl, err = tmpl.Parse(templates.Layout)
+	checkError(err)
+	tmpl, err = tmpl.Parse(templates.Form)
+	checkError(err)
+
+	data := make(map[string]interface{})
+	elements := make([]Element, len(srv.form))
+	copy(elements, srv.form)
```
Comment (mpsonntag) [inline tonic/routes.go:135]: Could not figure out how or where `srv.form` is actually set. `tonic` has a `SetForm` method but I could not find it being used anywhere.

File: tonic/routes.go [hunk 3]
```diff
@@ -0,0 +1,161 @@
...
+			return
+		}
+	} else {
+		userToken = tokens[0].Sha1
+	}
+
+	client = gogs.NewClient(srv.config.GINServer, userToken)
+	orgs, err := client.ListMyOrgs()
+
+	for idx, org := range orgs {
```
Comment (mpsonntag) [inline tonic/routes.go:77]: Is this missing a ToDo? It does not seem to be used for anything yet.

File: tonic/routes.go [hunk 4]
```diff
@@ -0,0 +1,161 @@
...
+		if err != nil {
+			srv.web.ErrorResponse(w, http.StatusUnauthorized, "authentication failed")
+			return
+		}
+	} else {
+		userToken = tokens[0].Sha1
+	}
+
+	client = gogs.NewClient(srv.config.GINServer, userToken)
+	orgs, err := client.ListMyOrgs()
```
[the remark this answers is not in the document]
Comment (achilleas-k) [reply tonic/routes.go:75]: This org listing and the loop below are leftovers from some testing.  I'll remove it.  It'll be back with the service that will be implemented to manage org repo creation.

File: tonic/routes.go [hunk 5]
```diff
@@ -0,0 +1,161 @@
...
+	if err := tmpl.Execute(w, data); err != nil {
+		log.Printf("Failed to render form: %v", err)
+	}
+}
+
+func (srv *Tonic) showJob(w http.ResponseWriter, r *http.Request) {
+	vars := mux.Vars(r)
+	jobid, err := strconv.ParseInt(vars["id"], 10, 64)
+	if err != nil {
+		srv.web.ErrorResponse(w, http.StatusInternalServerError, "Invalid ID")
```
[the remark this answers is not in the document]
Comment (achilleas-k) [reply tonic/routes.go:118]: Yup missing a return.  Thanks for catching.

File: tonic/routes.go (hunk 2 above)
Comment (achilleas-k) [reply tonic/routes.go:135]: Each instance that uses the library is responsible for defining its own form.  Basically, in its current state, the library is supposed to make it easy to define a form and a function that uses the data from that form.  The idea is that with those two elements you can build a bunch of different services that ask users for arbitrary input and automate actions on GIN on behalf of the user.

File: tonic/db/job.go (hunk 1 above)
Comment (achilleas-k) [reply tonic/db/job.go:52]: Yeah, that's a mistake.

**Pull request merged**